// log.io/lib/inputs.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getInputName = getInputName;
exports["default"] = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

/**
 * Get string name of provided input
 */
function getInputName(streamName, sourceName) {
  return "".concat(streamName, "|").concat(sourceName);
}
/**
 * Maintains list of registered inputs
 */


var InputRegistry = /*#__PURE__*/function () {
  function InputRegistry() {
    (0, _classCallCheck2["default"])(this, InputRegistry);
    (0, _defineProperty2["default"])(this, "inputs", void 0);
    this.inputs = {};
  }

  (0, _createClass2["default"])(InputRegistry, [{
    key: "add",
    value: function add(streamName, sourceName) {
      var inputName = getInputName(streamName, sourceName);
      this.inputs[inputName] = true;
      return inputName;
    }
  }, {
    key: "remove",
    value: function remove(streamName, sourceName) {
      var inputName = getInputName(streamName, sourceName);
      delete this.inputs[inputName];
      return inputName;
    }
  }, {
    key: "getInputs",
    value: function getInputs() {
      return Object.keys(this.inputs).map(function (inputName) {
        var _inputName$split = inputName.split('|'),
            _inputName$split2 = (0, _slicedToArray2["default"])(_inputName$split, 2),
            stream = _inputName$split2[0],
            source = _inputName$split2[1];

        return {
          stream: stream,
          source: source,
          inputName: inputName
        };
      });
    }
  }]);
  return InputRegistry;
}();

var _default = InputRegistry;
exports["default"] = _default;
