// log.io/lib/server.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _express = _interopRequireDefault(require("express"));

var _http = _interopRequireDefault(require("http"));

var _net = _interopRequireDefault(require("net"));

var _path = _interopRequireDefault(require("path"));

var _socket = _interopRequireDefault(require("socket.io"));

var _inputs = _interopRequireDefault(require("./inputs"));

// File path to UI app build artifacts (static JS/CSS/HTML)
var UI_BUILD_PATH = process.env.LOGIO_SERVER_UI_BUILD_PATH || _path["default"].resolve(__dirname, 'ui');
/**
 * Broadcast an inbound message to socket.io channels
 */


function handleNewMessage(_x, _x2, _x3, _x4) {
  return _handleNewMessage.apply(this, arguments);
}
/**
 * Broadcast a new input coming online to all browsers
 */


function _handleNewMessage() {
  _handleNewMessage = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(config, inputs, io, msgParts) {
    var _msgParts$slice, _msgParts$slice2, mtype, stream, source, msg, inputName;

    return _regenerator["default"].wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            _msgParts$slice = msgParts.slice(0, 3), _msgParts$slice2 = (0, _slicedToArray2["default"])(_msgParts$slice, 3), mtype = _msgParts$slice2[0], stream = _msgParts$slice2[1], source = _msgParts$slice2[2];
            msg = msgParts.slice(3).join('|');
            inputName = inputs.add(stream, source); // Broadcast message to input channel

            io.to(inputName).emit(mtype, {
              inputName: inputName,
              msg: msg,
              stream: stream,
              source: source
            }); // Broadcast ping to all browsers

            io.emit('+ping', {
              inputName: inputName,
              stream: stream,
              source: source
            });

            if (config.debug) {
              // eslint-disable-next-line no-console
              console.log(msgParts.join('|'));
            }

          case 6:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));
  return _handleNewMessage.apply(this, arguments);
}

function handleRegisterInput(_x5, _x6, _x7, _x8) {
  return _handleRegisterInput.apply(this, arguments);
}
/**
 * Broadcast an input going offline to all browsers
 */


function _handleRegisterInput() {
  _handleRegisterInput = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2(config, inputs, io, msgParts) {
    var _msgParts$slice3, _msgParts$slice4, mtype, stream, source, inputName;

    return _regenerator["default"].wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            _msgParts$slice3 = msgParts.slice(0, 3), _msgParts$slice4 = (0, _slicedToArray2["default"])(_msgParts$slice3, 3), mtype = _msgParts$slice4[0], stream = _msgParts$slice4[1], source = _msgParts$slice4[2];
            inputName = inputs.add(stream, source);
            io.emit(mtype, {
              stream: stream,
              source: source,
              inputName: inputName
            });

          case 3:
          case "end":
            return _context2.stop();
        }
      }
    }, _callee2);
  }));
  return _handleRegisterInput.apply(this, arguments);
}

function handleDeregisterInput(_x9, _x10, _x11, _x12) {
  return _handleDeregisterInput.apply(this, arguments);
} // Maps TCP message prefix to handler function


function _handleDeregisterInput() {
  _handleDeregisterInput = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3(config, inputs, io, msgParts) {
    var _msgParts$slice5, _msgParts$slice6, mtype, stream, source, inputName;

    return _regenerator["default"].wrap(function _callee3$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            _msgParts$slice5 = msgParts.slice(0, 3), _msgParts$slice6 = (0, _slicedToArray2["default"])(_msgParts$slice5, 3), mtype = _msgParts$slice6[0], stream = _msgParts$slice6[1], source = _msgParts$slice6[2];
            inputName = inputs.remove(stream, source);
            io.emit(mtype, {
              stream: stream,
              source: source,
              inputName: inputName
            });

          case 3:
          case "end":
            return _context3.stop();
        }
      }
    }, _callee3);
  }));
  return _handleDeregisterInput.apply(this, arguments);
}

var messageHandlers = {
  '+msg': handleNewMessage,
  '+input': handleRegisterInput,
  '-input': handleDeregisterInput
};
/**
 * Broadcast an inbound message to socket.io channels
 */

function broadcastMessage(_x13, _x14, _x15, _x16) {
  return _broadcastMessage.apply(this, arguments);
}
/**
 * Start message & web servers
 */


function _broadcastMessage() {
  _broadcastMessage = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee5(config, inputs, io, data) {
    var msgs;
    return _regenerator["default"].wrap(function _callee5$(_context5) {
      while (1) {
        switch (_context5.prev = _context5.next) {
          case 0:
            // Parse raw message into parts
            // NOTE: After split on null termination character, last item will always
            // be either an empty string or a partial/incomplete message
            msgs = data.toString().split('\0').slice(0, -1).filter(function (msg) {
              return !!msg.trim();
            });
            msgs.forEach( /*#__PURE__*/function () {
              var _ref = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee4(msg) {
                var msgParts, messageHandler;
                return _regenerator["default"].wrap(function _callee4$(_context4) {
                  while (1) {
                    switch (_context4.prev = _context4.next) {
                      case 0:
                        msgParts = msg.split('|');
                        messageHandler = messageHandlers[msgParts[0]];

                        if (!messageHandler) {
                          _context4.next = 7;
                          break;
                        }

                        _context4.next = 5;
                        return messageHandler(config, inputs, io, msgParts);

                      case 5:
                        _context4.next = 8;
                        break;

                      case 7:
                        // eslint-disable-next-line no-console
                        console.error("Unknown message type: ".concat(msgParts[0]));

                      case 8:
                      case "end":
                        return _context4.stop();
                    }
                  }
                }, _callee4);
              }));

              return function (_x18) {
                return _ref.apply(this, arguments);
              };
            }());

          case 2:
          case "end":
            return _context5.stop();
        }
      }
    }, _callee5);
  }));
  return _broadcastMessage.apply(this, arguments);
}

function main(_x17) {
  return _main.apply(this, arguments);
}

function _main() {
  _main = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee9(config) {
    var server, httpServer, io, inputs, messageServer;
    return _regenerator["default"].wrap(function _callee9$(_context9) {
      while (1) {
        switch (_context9.prev = _context9.next) {
          case 0:
            // Create HTTP server w/ static file serving & socket.io bindings
            server = (0, _express["default"])();
            server.use('/', _express["default"]["static"](UI_BUILD_PATH));
            httpServer = new _http["default"].Server(server);
            io = (0, _socket["default"])(httpServer);
            inputs = new _inputs["default"](); // Create TCP message server

            messageServer = _net["default"].createServer( /*#__PURE__*/function () {
              var _ref2 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee7(socket) {
                return _regenerator["default"].wrap(function _callee7$(_context7) {
                  while (1) {
                    switch (_context7.prev = _context7.next) {
                      case 0:
                        socket.on('data', /*#__PURE__*/function () {
                          var _ref3 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee6(data) {
                            return _regenerator["default"].wrap(function _callee6$(_context6) {
                              while (1) {
                                switch (_context6.prev = _context6.next) {
                                  case 0:
                                    _context6.next = 2;
                                    return broadcastMessage(config, inputs, io, data);

                                  case 2:
                                  case "end":
                                    return _context6.stop();
                                }
                              }
                            }, _callee6);
                          }));

                          return function (_x20) {
                            return _ref3.apply(this, arguments);
                          };
                        }());

                      case 1:
                      case "end":
                        return _context7.stop();
                    }
                  }
                }, _callee7);
              }));

              return function (_x19) {
                return _ref2.apply(this, arguments);
              };
            }()); // When a new browser connects, register stream activation events

            io.on('connection', /*#__PURE__*/function () {
              var _ref4 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee8(socket) {
                return _regenerator["default"].wrap(function _callee8$(_context8) {
                  while (1) {
                    switch (_context8.prev = _context8.next) {
                      case 0:
                        // Send existing inputs to browser
                        inputs.getInputs().forEach(function (input) {
                          socket.emit('+input', input);
                        }); // Register input activation events

                        socket.on('+activate', function (inputName) {
                          socket.join(inputName);
                        });
                        socket.on('-activate', function (inputName) {
                          socket.leave(inputName);
                        });

                      case 3:
                      case "end":
                        return _context8.stop();
                    }
                  }
                }, _callee8);
              }));

              return function (_x21) {
                return _ref4.apply(this, arguments);
              };
            }()); // Start listening for requests

            messageServer.listen(config.messageServer.port, config.messageServer.host, function () {
              // eslint-disable-next-line no-console
              console.log("TCP message server listening on port ".concat(config.messageServer.port));
            });
            httpServer.listen(config.httpServer.port, config.httpServer.host, function () {
              // eslint-disable-next-line no-console
              console.log("HTTP server listening on port ".concat(config.httpServer.port));
            });

          case 9:
          case "end":
            return _context9.stop();
        }
      }
    }, _callee9);
  }));
  return _main.apply(this, arguments);
}

var _default = main;
exports["default"] = _default;
